_posts/2015-09-02-counting-active-record-objects.md

```diff
@@ -0,0 +1,84 @@
+---
+layout: post
+title: Counting ActiveRecord Objects
+category: code
+---
+Recently, I found myself needing to DISTINCT a collection of records by a column, and then get a count of the filtered records. I also needed the records later on for more filtering, so I needed the full records rather than a single column. It sounds like it should be pretty straightforward:
+
+# The Scenario
+
+```
+# Halfway there
+Post.where(type: "code").count
+=> 6
+
+# The count I need to end up with, but with full records
+Post.where(type: "code").map(&:user_id).uniq.count
+=> 3
+```
+
+I have a Post class which has a "type" string column, a "user_id", and a "posted_on" date. I need to find posts with type "code", distinct them on the column user_id, and get the count of that, all the while keeping the collection of records around for further filtering on the posted_on date. I started with some straightforward queries:
+
+```
+Post.where(type: "code").select("distinct on (user_id) *")
+=> [#<Post:0x007fa0cd7f0538 id: 1, ...>,
+  #<Post:0x007fa0cd7f03d0 id: 2, ...>,
+  #<Post:0x007fa0cd7f0268 id: 3, ...>]
+```
+
+Great! I see three posts, which is the count that I need, and they all look correct. Now to just grab the count...
+
+```
+Post.where(type: "code").select("distinct on (user_id) *").count
+ERROR:  syntax error at or near "on"
+LINE 1: SELECT COUNT(distinct on (user_id) *) FROM "posts"
+        WHERE "po..."
+```
+
+Ok, so the count and 'distinct on' SQL don't play well with each other. Let's just use a different method to get the count.
+
+```
+Post.where(type: "code").select("distinct on (user_id) *").size
+=> 6
+```
+
+Well, we got an integer back this time, but it's the wrong number.
+We can see that the query returns only 3 records, but calling '.size' on it still returns '6' for some reason.
+
+Finally, in a moment of desperation, we fall back on another counting method:
+
+```
+Post.where(type: "code").select("distinct on (user_id) *").length
+=> 3
+```
+
+And we get the right number!
+
+# So what's going on here?
```

Comment:
i like how you do the thing then go on to explain how you do it. easy to follow, even for a pleab like me :P